### PR TITLE
[IMP] function: `hyperlink(url, label)`

### PR DESCRIPTION
**Pull request by yunchen4**
## Description:

This commit implements HYPERLINK() function which also exists in GSheets and Excel. 

The evaluated result is a markdown string, which will be evaluated as a link later. It behaves like inserting a link using menu, so not exactly the same as `HYPERLINK` in GSheets and Excel, which also supports something like `ftp` or `mailto`. 

Odoo task ID : [2837746](https://www.odoo.com/web#id=2837746&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: tests/functions/module_custom.test.ts [hunk 1]
```diff
@@ -130,3 +130,15 @@ describe("FORMAT.LARGE.NUMBER formula", () => {
     expect(evaluateCellText("A1", { A1: '=FORMAT.LARGE.NUMBER(100, "something")' })).toBe("#ERROR");
   });
 });
+
+describe("HYPERLINK formula", () => {
```
Comment (LucasLefevre) [inline tests/functions/module_custom.test.ts:134]: move to a new file `module_web.test.ts`

File: tests/functions/module_custom.test.ts [hunk 2]
```diff
@@ -130,3 +130,15 @@ describe("FORMAT.LARGE.NUMBER formula", () => {
...
+  test("The evaluated result is the markdown format of link", () => {
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("https://www.odoo.com", "Odoo")' })).toBe(
+      "Odoo"
+    );
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("https://www.odoo.com")' })).toBe(
+      "https://www.odoo.com"
+    );
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("invalidUrl")' })).toBe("invalidUrl");
```
Comment (LucasLefevre) [inline tests/functions/module_custom.test.ts:142]: test with
- reference `evaluateCell("A1", { A1: '=HYPERLINK(A2)', A2: "https://www.odoo.com" })`
- empty string `evaluateCellText("A1", { A1: '=HYPERLINK("")' })` and `evaluateCellText("A1", { A1: '=HYPERLINK("  ")' })`
- a number `evaluateCellText("A1", { A1: '=HYPERLINK(2)' })`
- a boolean `evaluateCellText("A1", { A1: '=HYPERLINK(true)' })`
- a date `evaluateCellText("A1", { A1: '=HYPERLINK("1/31/2022")' })`
- no argument `evaluateCellText("A1", { A1: '=HYPERLINK()' })`

File: tests/plugins/cell.test.ts [hunk 3]
```diff
@@ -64,23 +64,38 @@ describe("link cell", () => {
...
-      const cell = getEvaluatedCell(model, "A1");
+      let cell = getEvaluatedCell(model, "A1");
       expect(cell.link?.label).toBe("my label");
       expect(cell.link?.url).toBe(url);
       expect(urlRepresentation(cell.link!, model.getters)).toBe(url);
       expect(getCell(model, "A1")?.content).toBe(`[my label](${url})`);
       expect(getStyle(model, "A1")).toEqual({ textColor: LINK_COLOR });
       expect(getCellText(model, "A1")).toBe("my label");
+
+      setCellContent(model, "B1", `=HYPERLINK("${url}", "Odoo")`);
```
Comment (LucasLefevre) [inline tests/plugins/cell.test.ts:75]: move to another test. (same for the other tests)
Tests should be small and test one thing only.
This existing one tests markdown content, yours test HYPERLINK content

File: src/functions/module_custom.ts [hunk 4]
```diff
@@ -43,3 +43,23 @@ export const FORMAT_LARGE_NUMBER: AddFunctionDescription = {
...
+      "The text to display in the cell, enclosed in quotation marks."
+    )}
+  `),
+  returns: ["STRING"],
+  compute: function (url: PrimitiveArgValue, linkLabel: PrimitiveArgValue): string {
+    const processedUrl = toString(url);
+    const processedLabel = toString(linkLabel) || toString(url);
+    return toString(`[${processedLabel}](${processedUrl})`);
+  },
+  isExported: true,
```
Comment (LucasLefevre) [inline src/functions/module_custom.ts:64]: add the function to the exported function xlsx test here https://github.com/odoo/o-spreadsheet/blob/16.0/tests/xlsx_export.test.ts#L115

File: tests/functions/module_web.test.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    expect(
+      evaluateCellText("A1", { A1: '=HYPERLINK(A2, "Odoo")', A2: "https://www.odoo.com" })
+    ).toBe("Odoo");
+    expect(
+      evaluateCellText("A1", { A1: "=HYPERLINK(A2, A3)", A2: "https://www.odoo.com", A3: "Odoo" })
+    ).toBe("Odoo");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("invalidUrl")' })).toBe("invalidUrl");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("")' })).toBe("[]()");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK(" ")' })).toBe(" ");
+    expect(evaluateCellText("A1", { A1: "=HYPERLINK(2)" })).toBe("2");
```
Comment (LucasLefevre) [inline tests/functions/module_web.test.ts:23]: actually, you should use `evaluateCell` instead of `evaluateCellText` to assert the value is the *number* 2, not the the string "2". 
Same for every assertion, except maybe the date where both should be checked (the value which is a number and the text value which should be formatted as a date)

Sorry to have misguided you

File: tests/functions/module_web.test.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    );
+    expect(
+      evaluateCellText("A1", { A1: '=HYPERLINK(A2, "Odoo")', A2: "https://www.odoo.com" })
+    ).toBe("Odoo");
+    expect(
+      evaluateCellText("A1", { A1: "=HYPERLINK(A2, A3)", A2: "https://www.odoo.com", A3: "Odoo" })
+    ).toBe("Odoo");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("invalidUrl")' })).toBe("invalidUrl");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("")' })).toBe("[]()");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK(" ")' })).toBe(" ");
```
Comment (LucasLefevre) [inline tests/functions/module_web.test.ts:22]: also test empty label

File: tests/functions/module_web.test.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      "https://www.odoo.com"
+    );
+    expect(
+      evaluateCellText("A1", { A1: '=HYPERLINK(A2, "Odoo")', A2: "https://www.odoo.com" })
+    ).toBe("Odoo");
+    expect(
+      evaluateCellText("A1", { A1: "=HYPERLINK(A2, A3)", A2: "https://www.odoo.com", A3: "Odoo" })
+    ).toBe("Odoo");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("invalidUrl")' })).toBe("invalidUrl");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("")' })).toBe("[]()");
```
Comment (LucasLefevre) [inline tests/functions/module_web.test.ts:21]: that's weird. Does that mean it will display an empty link component ?
It should probably not be a link and just left empty.
It's a good test to add in `cell.test.ts`: check that `cell.link` is undefined in those situations
same for the line below

File: tests/functions/module_web.test.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK(" ")' })).toBe(" ");
+    expect(evaluateCellText("A1", { A1: "=HYPERLINK(2)" })).toBe("2");
+    expect(evaluateCellText("A1", { A1: "=HYPERLINK(true)" })).toBe("TRUE");
+    expect(evaluateCellText("A1", { A1: '=HYPERLINK("1/31/2022")' })).toBe("1/31/2022");
```
Comment (LucasLefevre) [inline tests/functions/module_web.test.ts:25]: add the same tests with the *label* being a number, boolean, date, .... (and a normal url)

File: tests/plugins/cell.test.ts [hunk 9]
```diff
@@ -74,14 +74,41 @@ describe("link cell", () => {
...
+      expect(getStyle(model, "B1")).toEqual({ textColor: LINK_COLOR });
+      expect(getCellText(model, "B1")).toBe(`=HYPERLINK("${url}", "Odoo")`);
+    }
+  );
+
+  test.each(["[Odoo](odoo.com)", '=HYPERLINK("odoo.com", "Odoo")'])(
+    "https prefix is added if it's missing: %s",
+    (content) => {
+      const model = new Model();
+      setCellContent(model, "A1", `${content}`);
```
Comment (LucasLefevre) [inline tests/plugins/cell.test.ts:96]:
```suggestion
      setCellContent(model, "A1", content);
```

File: tests/plugins/cell.test.ts [hunk 10]
```diff
@@ -74,14 +74,41 @@ describe("link cell", () => {
...
+      const cell = getEvaluatedCell(model, "A1");
+      expect(cell.link?.url).toBe("https://odoo.com");
+      expect(urlRepresentation(cell.link!, model.getters)).toBe("https://odoo.com");
+    }
+  );
+
+  test.each(['=HYPERLINK("")', '=HYPERLINK("   ")'])(
```
Comment (LucasLefevre) [inline tests/plugins/cell.test.ts:103]:
```suggestion
  test.each(['=HYPERLINK("")', '=HYPERLINK("   ")', '=HYPERLINK("", " ")', '=HYPERLINK(" ", " ")', '=HYPERLINK("", "")'])(
```

**Review by LucasLefevre — COMMENTED** [review]
almost good to go :)

File: src/functions/module_web.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  args: args(`
+    url (string) ${_lt("The full URL of the link enclosed in quotation marks.")}
+    link_label (string, optional) ${_lt(
+      "The text to display in the cell, enclosed in quotation marks."
+    )}
+  `),
+  returns: ["STRING"],
+  compute: function (url: PrimitiveArgValue, linkLabel: PrimitiveArgValue): string {
+    const processedUrl = toString(url).trim();
+    const processedLabel = toString(linkLabel) || processedUrl;
```
[the remark this answers is not in the document]
Comment (yunchen4) [reply src/functions/module_web.ts:21]: > We should not create the link if the trimmed label is also empty, right ? (and return the original non-trimmed label as simple value)

@LucasLefevre IMHO what you said makes more sense, but in Google Sheets they still convert it to a link, so I copied this behavior.

[the remark this answers is not in the document]
Comment (LucasLefevre) [reply src/functions/module_web.ts:21]: `=HYPERLINK("", "  ")` doesn't seem to create a link in G Sheet

**Comment by LucasLefevre** [comment]
robodoo r+
